File: cryptography/src/main/scala/akka/stream/alpakka/cryptography/impl/CryptographicFlows.scala
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.stream.alpakka.cryptography.impl

import java.security.{PrivateKey, PublicKey}
import javax.crypto.{Cipher, SecretKey}

import akka.NotUsed
import akka.stream.scaladsl.Flow
import akka.util.ByteString

object CryptographicFlows {

  def symmetricEncryptionFlow(secretKey: SecretKey): Flow[ByteString, ByteString, NotUsed] =
    Flow[ByteString].statefulMapConcat { () =>
      val cipher = Cipher.getInstance(secretKey.getAlgorithm)
      cipher.init(Cipher.ENCRYPT_MODE, secretKey)

      in => {
        val encrypted = cipher.doFinal(in.toArray)

        List(ByteString(encrypted))
      }
    }



  def symmetricDecryption(secretKey: SecretKey)(toDecode: ByteString): ByteString = {
    val cipher = Cipher.getInstance(secretKey.getAlgorithm)
    cipher.init(Cipher.DECRYPT_MODE, secretKey)
    val decrypted = cipher.doFinal(toDecode.toArray)

    ByteString(decrypted)
  }

  def asymmetricEncryption(publicKey: PublicKey)(in: ByteString): ByteString = {
    val cipher = Cipher.getInstance(publicKey.getAlgorithm)
    cipher.init(Cipher.PUBLIC_KEY, publicKey)
    val encrypted = cipher.doFinal(in.toArray)

    ByteString(encrypted)
  }

  def asymmetricDecryption(privateKey: PrivateKey)(toDecode: ByteString): ByteString = {
    val cipher = Cipher.getInstance(privateKey.getAlgorithm)
    cipher.init(Cipher.DECRYPT_MODE, privateKey)
    val decrypted = cipher.doFinal(toDecode.toArray)

    ByteString(decrypted)
  }
}

File: cryptography/src/main/scala/akka/stream/alpakka/cryptography/javadsl/CryptographicFlows.java
/*
 * Copyright (C) 2016-2018 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.stream.alpakka.cryptography.javadsl;

import akka.NotUsed;
import akka.stream.alpakka.cryptography.scaladsl.CryptographicFlows$;
import akka.stream.javadsl.Flow;
import akka.util.ByteString;

import javax.crypto.SecretKey;
import java.security.PrivateKey;
import java.security.PublicKey;

public class CryptographicFlows {
    public static Flow<ByteString, ByteString, NotUsed> symmetricEncryption(SecretKey secretKey) {
        return CryptographicFlows$.MODULE$.symmetricEncryption(secretKey).asJava();
    }

    public static Flow<ByteString, ByteString, NotUsed> symmetricDecryption(SecretKey secretKey) {
        return CryptographicFlows$.MODULE$.symmetricDecryption(secretKey).asJava();
    }

    public static Flow<ByteString, ByteString, NotUsed> asymmetricEncryption(PublicKey publicKey) {
        return CryptographicFlows$.MODULE$.asymmetricEncryption(publicKey).asJava();
    }

    public static Flow<ByteString, ByteString, NotUsed> asymmetricDecryption(PrivateKey privateKey) {
        return CryptographicFlows$.MODULE$.asymmetricDecryption(privateKey).asJava();
    }
}
